mypyc/genops.py

```diff
@@ -1126,23 +1159,101 @@ def generate_attr_defaults(self, cdef: ClassDef) -> None:
+        # https://github.com/python/cpython/blob/3.7/Lib/dataclasses.py#L957
+        filler_doc_str = 'filler docstring for classes decorated with dataclass'
+        self.primitive_op(dict_set_item_op,
+                         [ir.type_dict, self.load_static_unicode('__doc__'),
+                          self.load_static_unicode(filler_doc_str)], line)
+
+        name = self.load_static_unicode(ir.name)
+        type_obj = self.primitive_op(type_object_op, [], line)
+        class_type_obj = self.py_call(type_obj, [name, ir.type_bases, ir.type_dict], line)
+        return class_type_obj
+
+    def load_decorated_class(self, cdef: ClassDef, type_obj: Value) -> Value:
+        """
+        Given a decorated ClassDef and a register containing a non-extension representation of the
+        ClassDef created via the type constructor, applies the corresponding decorator functions
+        on that decorated ClassDef and returns a register containing the decorated ClassDef.
+        """
+        decorators = cdef.decorators
+        dec_class = type_obj
+        for d in reversed(decorators):
+            decorator = d.accept(self)
+            assert isinstance(decorator, Value)
+            dec_class = self.py_call(decorator, [dec_class], dec_class.line)
+        return dec_class
+
+    def add_to_type_dict(self, ir: ClassIR, key: str, val: Value, line: int) -> None:
+        # Add an attribute entry into the class dict of a non-extension class.
+        assert ir.type_dict is not None
+        key_unicode = self.load_static_unicode(key)
+        self.primitive_op(dict_set_item_op, [ir.type_dict, key_unicode, val], line)
+
+    def add_non_ext_class_attr(self, lvalue: NameExpr,
+                               stmt: AssignmentStmt, class_ir: ClassIR) -> None:
+        # We populate __annotations__ because dataclasses uses it to determine
+        # which attributes to compute on.
+        # TODO: Maybe generate more precise types for annotations
+        assert class_ir.type_anns is not None
+
+        key = self.load_static_unicode(lvalue.name)
+        typ = self.primitive_op(type_object_op, [], stmt.line)
+        self.primitive_op(dict_set_item_op, [class_ir.type_anns, key, typ], stmt.line)
+
+        # Only add the attribute to the __dict__ if the assignment is of the form:
+        # x : type = value (don't add attributes of the form 'x : type' to the __dict__).
+        if not isinstance(stmt.rvalue, TempNode):
+            self.add_to_type_dict(class_ir, lvalue.name, self.accept(stmt.rvalue), stmt.line)
+
     def visit_class_def(self, cdef: ClassDef) -> None:
-        self.allocate_class(cdef)
+        ir = self.mapper.type_to_ir[cdef.info]
+        # Currently, we only create non-extension classes for classes that are
+        # decorated. Classes decorated with @trait do not apply here, and are
+        # handled in a different way.
+        if ir.is_non_ext:
+            ir.type_dict = self.primitive_op(new_dict_op, [], cdef.line)
```

Comment:
Oh yeah it looks like you don't need to stick these into an object at all, you can just keep them in locals and pass them around?